### ISSUE TITLE
account not being created

### ISSUE THREAD

**Issue by johnmrncc1701d** [kinecosystem/kin-sdk-unity#8]
I followed the instructions for creating an account but it is not working. It is giving me a pointer to an account but when I try to get the public address I get an empty string. And also the account is not saved so the next time I run it, it thinks there is not account yet. Here is my code snippet. Can you help with this? ty

        m_kinClient = new KinClient(Kin.Environment.Test, "1acd");
        if (!m_kinClient.HasAccount())
        {
            try
            {
                m_account = m_kinClient.AddAccount();
                //here account public address is ""
            }
            catch (KinException e)
            {
                m_lastException = e;
            }
        }
        else
        {
            m_account = m_kinClient.GetAccount(0);
        }

**Comment by hitwill** [comment]
Are you testing this in Unity editor, or on an Android phone/ Simulator? The classes don't work in the editor, and will just return a null.

**Comment by johnmrncc1701d** [comment]
Ok I ran it on an emulator. It now gives the public ID. However when I do a GetStatus call on the account, it returns Not Created. What might this be? ty

**Comment by hitwill** [comment]
This is correct - When you create an account, you are only creating a **local** account on the user's phone. (like a paper wallet). For the account to be visible on the blockchain (and for GetStatus to return that the account exists **on the blockchain**) - you also have to create the account on the blockchain.


The sample code in Kin Onboarding class has shows you how to do it, but it's essentially this:


`


      public static IEnumerator CreateAccount(string publicAddress, Action<bool> onComplete = null)
         
      {
          var url = "http://friendbot-testnet.kininfrastructure.com/?addr=" + publicAddress;
          var req = UnityWebRequest.Get(url);
          yield return req.SendWebRequest();
            if (req.isNetworkError || req.isHttpError)
            {
                Debug.Log(req.error);
                if (onComplete != null)
                    onComplete(false);
            }
            else
            {
                if (onComplete != null)
                    onComplete(true);
            }
        }
`
So just pass your **local** address to this function, and it will use the friendbot to send a transaction on the blockchain to that address - that will make the account visible online as well.


When you switch to production mode, you will need to replace the url of the friendbot with the url of your own server, to fund the account with real Kin. The friendbot funds the account with test Kin on the test blockchain.

**Comment by johnmrncc1701d** [comment]
So when it is time to use my server in production mode to create the account on the blockchain, how do I have my server do that? I am thinking of just having the public address sent to my server even on test and have my server send it to the friendbot so I can swap it out in production easily. Is that better? ty

**Comment by hitwill** [comment]
Creating the account on the blockchain is honestly just sending some Kin to the address. (That's all!!) The Friendbot is just saving you time by sending test Kin to the address. If you change the friendbot url to your server, it will depend on what language you  are using.


I personally use node.js, so I use the [js-stellar-sdk](https://www.stellar.org/developers/js-stellar-sdk/reference/index.html).  So just pick a language you like from there, and use it to communicate with the Kin blockchain.


Makes sense?

**Comment by johnmrncc1701d** [comment]
yes that makes sense. ty

**Comment by johnmrncc1701d** [comment]
Ok I have another problem. I get the following error after making this call in Python. Do you know what the problem might be?

account.send_kin(customer_public_address, amount, fee=0, memo_text=description)

send transaction error with channel GAZVAGYNBBYEFSPNHIX3YE37I3ALMZYDYHN52P3IU6TXLZMKTP2YQB5E:   _data='{'type': 'transaction_failed', 'detail': 'The transaction failed when submitted to the stellar network. The `extras.result_codes` field on this response contains further details.  Descriptions of each code can be found at: https://www.stellar.org/developers/learn/concepts/list-of-operations.html', 'instance': None, 'title': 'Transaction Failed', 'extras':    _data='{'result_xdr': 'AAAAAAAAAAD/////AAAAAQAAAAAAAAAB////+wAAAAA=', 'invalid_field': None, 'result_codes':    _data='{'operations': ['op_no_destination'], 'transaction': 'tx_failed'}', 'envelope_xdr': 'AAAAADNQGw0IcELJ7TovvBN/RsC2ZwPB290/aKendeWKm/WIAAAAAAAPcXwAAAACAAAAAAAAAAEAAAARMS0wOEtjLUZvdW5kIEtpbiEAAAAAAAABAAAAAQAAAAAzUBsNCHBCye06L7wTf0bAtmcDwdvdP2inp3Xlipv1iAAAAAEAAAAAN5a/ka0DKuFOC6dV8Ii6YxhZEOzdB9UHoqqu/+Qju2sAAAAAAAAAAAABhqAAAAAAAAAAAYqb9YgAAABAb9qVzx5V+LO3TppU6NPsr/STJ1aJfFsY8jPeDaWxXYkLLzayhzrND4aIGYNrua/4x6rQRtXaQbmv1h0IfATODQ=='}', 'status': 400}'
127.0.0.1 - - [23/Mar/2019 19:30:03] "POST /earn HTTP/1.1" 500 -
Traceback (most recent call last):
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 2295, in wsgi_app
    response = self.handle_exception(e)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 1741, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\_compat.py", line 35, in reraise
    raise value
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\_compat.py", line 35, in reraise
    raise value
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "C:\Users\johnm\Anaconda3\lib\site-packages\flask\app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "D:\GameProjects\unity_workspace\kinprojects\DungeonParty\Server\server.py", line 81, in get_earn
    account.send_kin(client_transaction, amount, fee=0, memo_text=description)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\kin\account.py", line 192, in send_kin
    return self.submit_transaction(builder)
  File "C:\Users\johnm\Anaconda3\lib\site-packages\kin\account.py", line 274, in submit_transaction
    raise KinErrors.translate_error(e)
kin.errors.AccountNotFoundError:
        message='account not found'
        error_code='op_no_destination'
        extra data:

**Comment by hitwill** [comment]
Hard to tell without seeing the rest of the code - but I noticed you put 0 for the fee. Did you whitelist the transaction before sending?

**Comment by johnmrncc1701d** [comment]
Should it be whitelisted? My own account that the kin is sent from is whitelisted.

**Comment by hitwill** [comment]
Yes - that is at least one of the problems. If you set the fee to zero, you need to pass the transaction to a function that confirms the sender is whitelisted.

![image](https://user-images.githubusercontent.com/2501079/54996248-bc8d1f80-5003-11e9-8526-beab601b19ee.png)

After confirmation, it returns a callback, which then completes the transaction
![image](https://user-images.githubusercontent.com/2501079/54996303-e6464680-5003-11e9-86c6-c0e946fcb729.png)


So in summary, if you want 0 fees, you need to use 2 special functions:
1. Whitelist the transaction (confirm the account is whitelisted on Kin's servers)
2. Get the callback and complete the transaction

So it doesn't matter if your account is whitelisted on Kin's server - each *transaction* must be passed to Kin's servers, returned to you whitelisted, then you pass it on to the blockchain.

**Comment by johnmrncc1701d** [comment]
Is the fee being charged to the sender or receiver or is it taken from the amount sent? What if I want to send 1 Kin?

Also I read this from the Python SDK
"# By defualt, any payment sent from you is already considered whitelisted,
# so there is no need for this step for the server transactions"
So, since I am doing this on my Python server, I shouldn't need to whitelist my transaction.
I found this here: https://github.com/kinecosystem/kin-sdk-python/tree/v2-master#initialization

**Comment by hitwill** [comment]
Yes, confirmed from the documentation.  @Ronserruya - Any ideas on this issue?

**Comment by Ronserruya** [comment]
Hey @johnmrncc1701d , just like @hitwill said, account must be created first.
So you first need to use the python-sdk "create_account" method, and only then the "send_kin" method.

Regarding whitelisting, your account must be whitelisted by kin.
The sentence ""# By default, any payment sent from you is already considered whitelisted," was meant to say that if your server account is whitelisted, you dont need to do anything extra for your tx to be whitelisted"
I can see why is this confusing so I will change that.

Do you mind if move this issue to the python-sdk repo?

**Comment by johnmrncc1701d** [comment]
Sure you my move it. Just give me a link to the new location. 
How can I check if an account was already created from its public address?
ty
